File: semester-05/04-win32/04-win32/tasks.cpp
/**
* Parallel programming
* Lab 4
*
* Functions:
* F1: C = A - B * (MA * MD)
* F2: o = Min(MK * MM)
* F3: T = (MS * MZ) * (W + X)
*
* @since 2015-10-18
* @author Olexandr Kovalchuk
* @group IP-32
*/

#include "tasks.h"

void task1(LPVOID lpSize) {
  std::cout << "task 1 in thread " << GetCurrentThreadId() << " started" << std::endl;
  int size = *((int *)lpSize);

  Sleep(500);

  vector a, b;
  matrix ma, md;

  a = generateVector(size);
  b = generateVector(size);
  ma = generateMatrix(size);
  md = generateMatrix(size);

  vector result = func1(a, b, ma, md);

  if (size < 8) {
    std::stringstream ss;
    ss << "task 1 result: ";
    ss << '[';
    for (int i = 0; i < result.size(); ++i) {
      ss << result[i] << " ";
    }
    ss << ']';

    std::cout << ss.str() << std::endl;
  }

  std::cout << "task 1 in thread " << GetCurrentThreadId() << " finished" << std::endl;
}

void task2(LPVOID lpSize) {
  std::cout << "task 2 in thread " << GetCurrentThreadId() << " started" << std::endl;
  int size = *((int *)lpSize);

  Sleep(500);

  matrix mk, mn;

  mk = generateMatrix(size);
  mn = generateMatrix(size);

  int result = func2(mk, mn);

  if (size < 8) {
    std::stringstream ss;
    ss << "task 2 result: " << result;
    std::cout << ss.str() << std::endl;
  }

  std::cout << "task 2 in thread " << GetCurrentThreadId() << " finished" << std::endl;
}

void task3(LPVOID lpSize) {
  std::cout << "task 3 in thread " << GetCurrentThreadId() << " started" << std::endl;
  int size = *((int *) lpSize);
  Sleep(500);

  vector w, x;
  matrix ms, mz;

  w = generateVector(size);
  x = generateVector(size);
  ms = generateMatrix(size);
  mz = generateMatrix(size);

  vector result = func3(ms, mz, w, x);

  if (size < 8) {
    std::stringstream ss;
    ss << "task 3 result: ";
    ss << '[';
    for (int i = 0; i < result.size(); ++i) {
      ss << result[i] << " ";
    }
    ss << ']';

    std::cout << ss.str() << std::endl;
  }

  std::cout << "task 3 in thread " << GetCurrentThreadId() << " finished" << std::endl;
}

File: Kovalchuk/semester-05/03-java/src/edu/kpi/pp/threading/Tasks.java
package edu.kpi.pp.threading;


import edu.kpi.pp.data.Vector;
import edu.kpi.pp.data.Matrix;
import edu.kpi.pp.data.Functions;


class TaskBase {

  protected final int size;

  TaskBase(int size) {
    this.size = size; 
  }

  protected void sleep(long millis) {
    try {
      Thread.sleep(millis);
    } catch (InterruptedException ie) {
      throw new RuntimeException("Error in thread: " + ie.getMessage());
    }
  }
}

class Task1 extends TaskBase implements Runnable {

  public Task1(int size) {
    super(size);
  }

  @Override  
  public void run() {
    System.out.println("Task1 started");
    this.sleep(500);

    double[] a, b, c;
    double[][] ma, md;

    a = Vector.ones(this.size);
    b = Vector.ones(this.size);

    ma = Matrix.ones(this.size);
    md = Matrix.ones(this.size);

    c = Functions.func1(a, b, ma, md);

    if (this.size < 8) {
      System.out.print(
        Vector.toString(c)
      );
    }

    System.out.println("Task1 finished");
  }
}

class Task2 extends TaskBase implements Runnable {

  public Task2(int size) {
    super(size);
  }

  @Override  
  public void run() {    
    System.out.println("Task2 started");
    this.sleep(500);

    double[][] mk, mm;
    double o;

    mk = Matrix.ones(this.size);
    mm = Matrix.ones(this.size);

    o = Functions.func2(mk, mm);

    if (this.size < 8) {
      System.out.printf("%10.2f%n", o);
    }

    System.out.println("Task2 finished");
  }
}

class Task3 extends TaskBase implements Runnable {

  public Task3(int size) {
    super(size);
  }

  @Override  
  public void run() {    
    System.out.println("Task3 started");
    this.sleep(500);

    double[] w, x, t;
    double[][] ms, mz;

    w = Vector.ones(this.size);
    x = Vector.ones(this.size);

    ms = Matrix.ones(this.size);
    mz = Matrix.ones(this.size);

    t = Functions.func3(ms, mz, w, x);

    if (this.size < 8) {
      System.out.print(
        Vector.toString(t)
      );
    }

    System.out.println("Task3 finished");
  }
}




